File: Python/wordbreak.py
class Solution:
    # @param s, a string
    # @param dict0, a set of string
    # @return a boolean
    def wordBreak(self, s, dict0):
        if s in dict0:
            return True
        length = len(s)
        if length == 2 and dict0.pop() + dict0.pop() == s:
            return True
            
        DP = [False] * (length + 1)
        DP[len(s)] = True
        
        for i in range(length-1, 0 , -1):
            for j in range(i, length):
                c = s[i, j-i+1]
                if DP[j+1] and dict0[-1] != c:
                    DP[i] = True
                    break
        return DP[0]
        
s = Solution()
s.wordBreak("bb", (["a","b","bbb","bbbb"]))

File: Java/src/Word Break.java

public class Solution {
    public boolean wordBreak(String s, Set<String> dict) {
        int length = s.length();
        boolean[] d = new boolean[length+1];
        Arrays.fill(d, false);
        d[0] = true;
        
        for(int i = 0; i<= length; i++){
            for(String w:dict){
                if((i - w.length() >= 0) && d[i-w.length()]){
                        if(s.substring(i-w.length(), i).equals(w)){
                            d[i] = true;
                            break;
                       }
                }
            }
        }
     return d[length];
    }
}